metadata/src/main/java/org/apache/kafka/controller/MetaLogManager.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.ApiMessageAndVersion;
+
+/**
+ * The MetaLogManager handles storing metadata and electing leaders.
+ */
+public interface MetaLogManager extends AutoCloseable {
+    /**
+     * Listeners receive notifications from the MetaLogManager.
+     */
+    interface Listener {
+        /**
+         * Called when the MetaLogManager commits a message.
+         *
+         * @param epoch         The controller epoch of the message.
+         * @param index         The index of the message within the controller epoch.
+         * @param message       The message.
+         */
+        void commit(long epoch, long index, ApiMessage message);
+
+        /**
+         * Called when the MetaLogManager has claimed the leadership.
+         *
+         * @param epoch         The controller epoch that is starting.
+         */
+        void claim(long epoch);
+
+        /**
+         * Called when the MetaLogManager has renounced the leadership.
+         *
+         * @param epoch         The controller epoch that has ended.
+         */
+        void renounce(long epoch);
+
+        /**
+         * Called when the MetaLogManager has finished shutting down, and wants to tell its
+         * listener that it is safe to shut down as well.
+         */
+        void beginShutdown();
```

Comment:
All the methods should be from the perspective of the listener.  They're all focused on responding to things that have happened in the log.  I can add "handle" before these methods if that makes it clearer.

I use `beginShutdown` rather than `shutdown` to indicate that the method isn't blocking.  It starts a shutdown, but doesn't wait for it to complete.

Reply:
> All the methods should be from the perspective of the listener.
> I can add "handle" before these methods if that makes it clearer.

Yeah, that would help (`handleClaim()`, `handleRenounce()`, etc.)

Reply:
done